pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/BrokerRegistryImpl.java

```diff
@@ -127,12 +140,35 @@ public CompletableFuture<Void> registerAsync() {
         }
         log.info("[{}] Started registering self to {} (state: {})", getBrokerId(), brokerIdKeyPath, state);
         return brokerLookupDataMetadataCache.put(brokerIdKeyPath, brokerLookupData, EnumSet.of(CreateOption.Ephemeral))
-                .thenAccept(__ -> {
-                    this.state.set(State.Registered);
-                    log.info("[{}] Finished registering self", getBrokerId());
+                .orTimeout(pulsar.getConfiguration().getMetadataStoreOperationTimeoutSeconds(), TimeUnit.SECONDS)
+                .whenComplete((__, ex) -> {
+                    if (ex == null) {
+                        this.state.set(State.Registered);
+                        log.info("[{}] Finished registering self", getBrokerId());
+                    } else {
+                        log.error("[{}] Failed registering self", getBrokerId(), ex);
```

Comment:
In this case (`ex != null`), the future of `registerAsync()` won't fail.

Please re-throw `ex` here, e.g.

```
                        throw new CompletionException(ex);
```

Or you can just print the error logs in `doRegisterAsyncWithRetries` and keep this method not changed

Reply:
> whenComplete has already handles the exception and it does not propagate the exception.

Sorry. Maybe I misunderstood this CompletableFuture.whenComplete behavior. Doesn't it propagate the previous stage's exception?

![Screen Shot 2024-10-02 at 11 02 53 PM](https://github.com/user-attachments/assets/5d5f0c11-6fc3-45f0-9085-2d570b1889b7)

Reply:
Oh I see. I've misunderstood the semantics of `whenComplete`.